---
title: "Hsp70 binding sites in the yeast proteome"
author: "D. Allan Drummond"
date: "5/25/2018"
output: html_document
---

output:
  html_document:
    toc: true
    toc_depth: 4
---



```{r setup,warning=FALSE,message=FALSE,echo=FALSE}

## knitr options for report generation
knitr::opts_chunk$set(warning=FALSE,message=FALSE,echo=TRUE,cache=TRUE,
                      results="show",
                      fig.path="figure/control-for-length-",
                      cache.path="cache/control-for-length-")


## data processing options, packages, functions
options(stringsAsFactors=FALSE)  # load character strings as strings
library(plyr)
library(reshape2) # data manipulation
library(tidyverse) # more data manipulation
library(stringr)
library(ggrepel)
#library(plotly) # interactive graphics
library(lmodel2)

library(cowplot) # extensions and nice defaults
 # set default theme for graphics
# theme_set(theme_bw() %+replace% theme(panel.grid.minor=element_blank())) 
theme_set(theme_cowplot(font_size=11) %+replace% 
              theme(# panel.grid.major=element_line(size=0.15,colour="grey50"),
                  panel.border=element_rect(colour = "grey50",
                                            linetype = "solid",size=0.5),
                    strip.background = element_blank()))

graycol <- "#333333cc"
orangecol <- "#cc5500cc"
bluecol <- "#0000aacc"
greencol <- "#22cc00cc"
purplecol <- "#cc22cccc"
cyancol <- "#2aa198cc"
redcol <- "#dc322fcc"
violetcol <- "#6c71c4cc"

```

```{r utility, child="../scripts_processing/utilityFunctions.Rmd"}
```

```{r load_data, echo=FALSE, warning=FALSE}
# setwd("~/Repos/HeatShockQmRNA/scripts_analysis/")
psup_levels <- c('30C','42C','42CR','46C')

h70 <- read_tsv("../data/scer-motifs-summary-r12_15.txt", comment='#')

smpae <- read_tsv("../../HeatShockQmRNA/src/scer-mrna-protein-absolute-estimate.txt",comment="#") %>%
    dplyr::rename(ORF=orf)
bg <- read_tsv("../../HeatShockQmRNA/index/scer_orf_transcript_length.txt") %>% left_join(smpae,by="ORF")

# Labeling
glyc <- read_tsv("../../HeatShockQmRNA/src/scer-glycolytic-proteins.txt", comment='#')
rib <- read_tsv("../../HeatShockQmRNA/src/scer-ribosomal-proteins.txt", comment='#')
#hs <- read_tsv("../src/scer-heat-shock-proteins.txt", comment='#')
hs <- read_tsv("../../HeatShockQmRNA/src/heat-shock-42C-up-genes-auto.txt", comment='#')
gtpase <- read_tsv("../../stress-ph/data/GTPase_activity_annotations.txt", comment='#') %>% mutate(orf=`Gene Systematic Name`)
agg <- read_tsv("../../HeatShockQmRNA/src/heat_aggregators.txt", comment='#')

gtpase.orf <- unique(gtpase$orf)
agg.gtpase <- na.omit(agg[match(gtpase.orf,agg$orf),])
agg.gtpase <- c('NOG1','SUP35','FUN12','NUG1','NOG2','LSG1','BMS1','DNM1','TUF1')
ctrl.gtpase <- c('ARF1','TSR1','EFT1','EFT2','RAS2','GSP1')

# fast
# 
# lsg1,nug1,nog2,nog1,bms1
# slow
# sup35,tuf1,dnm1,fun12
# no
# ARF1,TSR1,EFT1,EFT2,RAS2,GSP1

h70$label <- NA
h70[!is.na(match(h70$orf, agg$orf)),'label'] <- 'aggregator'
h70[!is.na(match(h70$orf, rib$orf)),'label'] <- 'ribosomal'
h70[!is.na(match(h70$orf, glyc$orf)),'label'] <- 'glycolytic'
h70[!is.na(match(h70$orf, hs$orf)),'label'] <- 'heat-shock'
h70[!is.na(match(h70$orf, subset(agg, superaggregator)$orf)),'label'] <- 'superaggregator'
#h70$label <- factor(h70$label, levels=names(sort(table(h70$label))))

h70 <- h70 %>% left_join(bg, by=c(orf='ORF'))
```

```{r weight_matrix, echo=FALSE, warning=FALSE}
mat <- read_tsv("../data/rudiger-hsp70-pssm.txt", comment='#')
matm <- mat %>% gather(res, score, -aa) %>% extract(res, "position", "p([[:alnum:]]+)")
matm$position <- factor(as.numeric(matm$position))
matm$aa <- factor(matm$aa, levels=strsplit("GAMVILFYWQNDESTPHRKC", split='')[[1]])
ggplot(matm, aes(x=position, y=aa, fill=score)) + geom_tile() + scale_fill_gradient2(low = bluecol, high = orangecol, mid = "white", midpoint = 0, space = "Lab", name="Binding score") + ylab("Amino acid") + xlab("Residue position in peptide")
```
```{r sequence_logo, dependson="weight_matrix", echo=FALSE, warning=FALSE}
library(ggseqlogo)
matl <- as.matrix(mat %>% select(-aa))
matl <- scale(exp(matl), center=FALSE, scale=colSums(exp(matl)))
rownames(matl) <- mat$aa
rudiger.logo <- ggseqlogo(matl)
ggdraw(rudiger.logo)
```


```{r categories, dependson="load_data", echo=FALSE, warning=FALSE}
score.thresh <- as.numeric(quantile(h70$max.score, probs=c(0.5)))
h70sub <- h70 %>% filter(!(label %in% c('glycolytic','ribosomal')))
h70sub[!is.na(match(h70sub$orf, subset(agg, superaggregator)$orf)),'label'] <- 'aggregator'
g <- ggplot(h70sub, aes(x=max.score, colour=label)) + stat_ecdf() + xlab('Maximum Hsp70 binding score') + ylab('Proportion')
ggdraw(g)

#tops <- h70 %>% filter(max.score>=score.thresh)
y <- h70 %>% filter(max.score>=score.thresh & label=='aggregator')
print(tops <- y[order(y$max.score,decreasing=T),c('orf','gene','max.score','num.motifs','label','mrna','prot')])
```

```{r profile, echo=FALSE, warning=FALSE, fig.asp=2, fig.width=4}

# https://github.com/stas-g/findPeaks/blob/master/find_peaks.R
find_peaks <- function (x, m = 3){
    shape <- diff(sign(diff(x, na.pad = FALSE)))
    pks <- sapply(which(shape < 0), FUN = function(i){
       z <- i - m + 1
       z <- ifelse(z > 0, z, 1)
       w <- i + m + 1
       w <- ifelse(w < length(x), w, length(x))
       if(all(x[c(z : i, (i + 2) : w)] <= x[i + 1])) return(i + 1) else return(numeric(0))
    })
     pks <- unlist(pks)
     pks
}

plot_profile <- function(gene, cutoff=10) {
  compare.set <- h70 %>% filter(label %in% c('aggregator','superaggregator'))
  x <- read_tsv(paste("../data/",tolower(gene),"-motifs-r10.txt",sep=''), comment='#')
  g <- ggplot(x, aes(pos, y=score)) + #geom_hline(yintercept = cutoff, linetype='dotted') + 
    geom_hline(yintercept=quantile(compare.set$max.score, probs=c(0.5,0.9,0.99),na.rm=T), colour=c(greencol,bluecol,orangecol)) +
    geom_line() + coord_cartesian(ylim=c(-30,30)) +
    xlab("Position (residue number)") + ylab("Hsp70 binding score") + ggtitle(gene)
  xv <- read_tsv(paste("../data/",tolower(gene),"-motifs-r10.txt",sep=''), comment='#')
  #g <- g + geom_line(data=xv, colour=graycol)
  #y <- x %>% filter(score >= cutoff)
  #if (nrow(y)>0) {
    keysite <-  x[find_peaks(x$score),] %>% filter(score>=cutoff)
    g <- g + geom_label_repel(data=keysite, aes(label=window, y=score), nudge_y=10, segment.color = 'red', segment.size=0.5, seed=111)
  #}
  g
}

cutoff <- as.numeric(quantile(h70$max.score, probs=c(0.5)))

hsf1 <- read_tsv("../data/hsf1-motifs-r10.txt", comment='#')
#cdc39 <- read_tsv("../data/cdc39-motifs-r10.txt", comment='#')
#nug1 <- read_tsv("../data/nug1-motifs-r10.txt", comment='#')
ggdraw(g <- plot_grid(
  ghsf <- plot_profile('hsf1', cutoff=cutoff), 
  gnug <- plot_profile('nug1', cutoff=cutoff), 
  gcdc <- plot_profile('nog1', cutoff=cutoff), 
  gcdc <- plot_profile('nog2', cutoff=cutoff), 
  ncol=1))
```

```{r full_distribution}
#d <- read_tsv("../data/scer-motifs-distribution.txt", comment='#')
x <- read_tsv("~/research/hsp70-binding/data/scer-motifs-summary-r12_15.txt", comment='#')
agg <- read_tsv("../data/heat_aggregators.txt", comment='#')

hsf1 <- read_tsv("../data/hsf1-motifs-r10.txt", comment='#')
nug1 <- read_tsv("../data/nug1-motifs-r10.txt", comment='#')

addAnnotation <- function(the_plot, gene, color, nudge) {
  x <- read_tsv(paste("../data/",gene,"-motifs-r10.txt",sep=''), comment='#')
  max.x <- x %>% filter(score==max(x$score))
  the_plot <- the_plot + geom_density(data=x, colour=color, kernel='r') + 
    geom_point(data=max.x, aes(y=0.005), shape=25, size=5, fill=color, colour=color) +
    geom_text_repel(data=max.x, aes(y=0.005, label=gene), colour=color, nudge_y=nudge)
  the_plot
}

g <- ggplot(d, aes(x=score, y=..density..)) + geom_density(kernel='r', size=1) + 
  geom_density(data=x, aes(x=max.score), kernel='r', colour='gray', linetype=2) + 
  scale_x_continuous(limits = c(-25,25))
g <- addAnnotation(g, "PAB1", "darkgreen", nudge=0)
g <- addAnnotation(g, "NUG1", purplecol, nudge=0.02)
g <- addAnnotation(g, "HSF1", orangecol, nudge=0.04)
g <- addAnnotation(g, "DED1", redcol, nudge=0.06)
g <- addAnnotation(g, "OLA1", violetcol, nudge=0.08)
print(g)
```





